src/node_api_helpers.h

```diff
@@ -123,6 +413,84 @@ namespace Napi {
     char str_st_[1024];
   };
 
+  class ObjectWrap {
+   public:
+    ObjectWrap() {
+      napi_env env = napi_get_current_env();
+      handle_ = napi_create_persistent(env, nullptr);
+      refs_ = 0;
+    }
+
+    virtual ~ObjectWrap() {
+      napi_env env = napi_get_current_env();
+      napi_release_persistent(env, persistent());
+    }
+
+    template <class T>
+    static inline T* Unwrap(napi_value object) {
+      napi_env env = napi_get_current_env();
+      assert(!napi_is_empty(env, object));
+      assert(napi_get_internal_field_count(env, object) > 0);
+      void* ptr = napi_get_internal_field_pointer(env, object, 0);
+      ObjectWrap* wrap = static_cast<ObjectWrap*>(ptr);
+      return static_cast<T*>(wrap);
+    }
+
+    inline napi_value handle() {
+      napi_env env = napi_get_current_env();
+      return napi_get_persistent_value(env, persistent());
+    }
+
+    inline napi_persistent persistent() {
+      return handle_;
+    }
+
+   protected:
+    inline void Wrap(napi_value object) {
+      napi_env env = napi_get_current_env();
+      assert(napi_is_persistent_empty(env, persistent()));
+      assert(napi_get_internal_field_count(env, object) > 0);
+      napi_set_internal_field_pointer(env, object, 0, this);
+      napi_release_persistent(env, persistent());
+      handle_ = napi_create_persistent(env, object);
+      napi_persistent_make_weak(env, persistent(), this);
+    }
+
+    /* Ref() marks the object as being attached to an event loop.
+     * Refed objects will not be garbage collected, even if
+     * all references are lost.
+     */
+    virtual void Ref() {
+      napi_env env = napi_get_current_env();
+      assert(!napi_is_persistent_empty(env, persistent()));
+      napi_persistent_clear_weak(env, persistent());
+      refs_++;
+    }
+
+    /* Unref() marks an object as detached from the event loop.  This is its
+     * default state.  When an object with a "weak" reference changes from
+     * attached to detached state it will be freed. Be careful not to access
+     * the object after making this call as it might be gone!
+     * (A "weak reference" means an object that only has a
+     * persistent handle.)
+     *
+     * DO NOT CALL THIS FROM DESTRUCTOR
+     */
+    virtual void Unref() {
+      napi_env env = napi_get_current_env();
+      assert(!napi_is_persistent_empty(env, persistent()));
+      assert(refs_ > 0);
+      if (--refs_ == 0)
+        napi_persistent_make_weak(env, persistent(), this);
+    }
+
+    int refs_;
+
+   private:
+    napi_persistent handle_;
+};
+
+
```

Comment:
We already have  C api to provide ObjectWrap functionality (as used in existing object wrap examples https://github.com/nodejs/abi-stable-node-addon-examples/blob/master/6_object_wrap/nan/myobject.h ) .  I think we should re-use that instead of re-creating. If there is something missing then we should add to that existing API.